Fix python variant of merge_config

nashif opened the pull request:
Do not use comment indicating variable is not set, this confuses the merge
script.

Fixes #5374

nashif committed:
kconfig: use merge_config.py for merging

Changed the python variant to support fragments with comments indicating
an option is not being set.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented:
Should we fix the parsing of these "not set" lines for the future, in case someone mistakenly adds one?

nashif commented:
> Should we fix the parsing of these "not set" lines for the future, in case someone mistakenly adds one?

I am still not sure if this was actually valid but we can do that, it will require rewriting the python script though.

mbolivar commented:
I think the "is not set" syntax is valid kconfig fragment syntax, even though it looks like a comment. Are you sure about this?

nashif commented:
> I think the "is not set" syntax is valid kconfig fragment syntax, even though it looks like a comment. Are you sure about this?

it is valid for kconfig in general, was not sure if it is supported as fragment. We have been using direct assignment instead with the exception of the few files changed in the PR. 
We do not need to support this I guess, this will require doing it in a fashion similar to how it is done in the shell script thus rewriting the python script.

SebastianBoe commented:
I believe we should support this "as fragment", because otherwise it wouldn't be valid
for a user to do

cp build/zephyr/.config prj.conf

which is the recommended way to save a GUI-edited .conf file isn't it?

dbkinder commented:
In the title, veriant should be variant.

carlescufi commented:
Is “is not set” used only for variables that default to “n”, and never for ones that have been explicitly ser to “n” but default to “y”? Maybe we could simply ignore lines that begin with “#”?

nashif commented:
I now have a version that mimics the original shell script and does exactly the same, will upload after more testing

nashif commented:
updated, still WIP, just uploaded to get through sanitycheck

nashif commented:
ok, I think i have it now, please review

SebastianBoe commented on scripts/kconfig/merge_config.py:
```diff
@@ -3,50 +3,92 @@
+    else:
+        with open('%s/.config' % args.output_dir, 'w') as file:
+            file.write(initfile_content)
+        with open('%s/.config.test' % args.output_dir, 'w') as file:
```
Is this debug code that was included by mistake?

nashif replied:
yep

SebastianBoe commented:
There is something off with the redundancy-detection logic. I have a bunch of fixes/refactorings on top of your work so I'll do a PR to your branch that fixes the fact that the redundancy detector uses substring to detect redundancies.

`            prev=list(filter(lambda x: cfg in x, initfile_content.split("\n")))
`


```
sebo@mach:~/zephyr/samples/net/irc_bot$ /usr/bin/python3 /home/sebo/zephyr/scripts/kconfig/merge_config.py -m -O /home/sebo/zephyr/samples/net/irc_bot/b/zephyr /home/sebo/zephyr/boards/riscv32/zedboard_pulpino/zedboard_pulpino_defconfig /home/sebo/zephyr/samples/net/irc_bot/prj.conf
-- Using /home/sebo/zephyr/boards/riscv32/zedboard_pulpino/zedboard_pulpino_defconfig as base
-- Merging /home/sebo/zephyr/samples/net/irc_bot/prj.conf
-- Value of CONFIG_RISCV is redefined by fragment /home/sebo/zephyr/samples/net/irc_bot/prj.conf
-- Previous value: CONFIG_RISCV32=y
-- New value: CONFIG_RISCV=y
sebo@mach:~/zephyr/samples/net/irc_bot$ git diff
WARNING: terminal is not fully functional
-  (press RETURN)
diff --git a/samples/net/irc_bot/prj.conf b/samples/net/irc_bot/prj.conf
index bc88318..0e73d90 100644
--- a/samples/net/irc_bot/prj.conf
+++ b/samples/net/irc_bot/prj.conf
@@ -1,5 +1,7 @@
 CONFIG_INIT_STACKS=y
 
+CONFIG_RISCV=y
+
 CONFIG_NET_LOG_ENABLED=y
 CONFIG_SYS_LOG_NET_LEVEL=4
 CONFIG_NET_IPV4=y
sebo@mach:~/zephyr/samples/net/irc_bot$ git log -1
WARNING: terminal is not fully functional
-  (press RETURN)
commit 7cbac50536880a15beb530f75242a3739ae9b978
Author: Anas Nashif <anas.nashif@intel.com>
Date:   Sun Dec 17 08:23:11 2017 -0500

    kconfig: use merge_config.py for merging
    
    Changed the python variant to support fragments with comments indicating
    an option is not being set.
    
    Signed-off-by: Anas Nashif <anas.nashif@intel.com>

```

nashif commented:
> prev=list(filter(lambda x: cfg in x, initfile_content.split("\n")))

maybe 

`prev=list(filter(lambda x: cfg in x.split(), initfile_content.split("\n")))`

SebastianBoe committed:
merge_config: Rewrite merge_config.py

Rewrite of merge_config.py, several unrelated changes including;

Do not attribute meaning to lines that look like this:

Only to lines that look like this:

Removed the unimplemented 'runmake' and 'alltarget' arguments.

Invoke the python interpreter instead of the script directly to avoid
reliance on bash's shebang feature.

Use exact line match instead of substring to detect redundant
config's.

And more ...

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
@nashif : Various fixes based on your work can be found here: https://github.com/zephyrproject-rtos/zephyr/issues/5444

Commit:
Merge pull request #2 from SebastianBoe/merge_config_rework

rewrite merge_config

mbolivar commented:
> @nashif : Various fixes based on your work can be found here: nashif#2

I've got a few changes I'd like to propose to the combined efforts so far; putting my review together now

lpereira commented:
As a nitpick: it's often better to use a list comprehension instead of `filter()`, specially if the result is going to be converted to a list afterwards:  `prev=list(filter(lambda x: cfg in x.split(), initfile_content.split("\n")))` can be written as `prev = [line for line in initfile_content.split("\n") if cfg in line.split()]`.

mbolivar commented on scripts/kconfig/merge_config.py:
```diff
@@ -3,50 +3,91 @@
+                        is_redefine  = line != prev_line
+                        if not args.quiet:
+                            if is_redefine:
+                                print("-- Value of {} is redefined by fragment {}".format(config_name, fragment_path))
```
I think it'd be an improvement to print these to stderr

mbolivar reviewed (changes requested):
I think it's almost there, but still has some issues. Here is an update which hopefully fixes the issues I've pointed out in this review:

https://github.com/mbolivar/zephyr/tree/merge_config-py

nashif commented:
who would have thought such a small script would take so much effort to convert... :)

mbolivar commented:
Right?!

mbolivar commented:
> https://github.com/mbolivar/zephyr/tree/merge_config-py

^^

@nashif @SebastianBoe just wanted to make sure you saw that I've got a patch which addresses the issues. No issues from me if you want to squash everything together (as long as SOB lines are preserved of course).

SebastianBoe commented on scripts/kconfig/merge_config.py:
```diff
@@ -3,50 +3,91 @@
+    # In other words [(config_name, config_line), ... ]
+    #
+    # Note that "# CONFIG_ARM is not set" is not the same as a
+    # comment, it has meaning (for now).
```
I disagree. IMHO "# is not set" should eventually be deprecated in favour of "=n" for obvious reasons.

Zephyr usability trumps Linux kernel Kconfig compatibility.

mbolivar replied:
> I disagree. IMHO "# is not set" should eventually be deprecated in favour of "=n" for obvious reasons.

I can see why you say so.

There are many more people who are familiar with the Linux style than the Zephyr style, though, and I think it's "obvious" that we want to obey principle of least surprise, especially since Zephyr is a Linux Foundation project.

@nashif, what's your decision?

carlescufi replied:
If I may weigh in, I think that both formats should be supported and should do exactly the same:
```
CONFIG_BT=n
# CONFIG_BT is not set
```
Should both turn Bluetooth OFF if present in the .conf or defconfig files

carlescufi replied:
Maybe I've missed something, but does `conf` treat those 2 differently in the `.config` file?

SebastianBoe replied:
To get this merged, I suggest we retain KBuild compatibilty for now, and then we can discuss the long-term Kconfig syntax at a later point.

SebastianBoe commented on scripts/kconfig/merge_config.py:
```diff
@@ -3,50 +3,91 @@
+    # state like any other
+
+    is_not_set_pattern = re.compile('^# (CONFIG_[a-zA-Z0-9_]*) is not set.*')
+    pattern            = re.compile('^(CONFIG_[a-zA-Z0-9_]*)[= ].*')
```
> "It matches strings like "CONFIG_FOO did not actually receive a value"

That is a minor problem because that is invalid syntax, but it would be great for
usability if you were able to fix this!

> If the base file sets "# CONFIG_FOO is not set", and some other fragment sets "CONFIG_FOO=n", that gets treated like an override, even though it's redundant (and CONFIG_FOO=n should be changed to "is not set").

I don't see how this affects the user, but if you can find a clean fix then why not :)

mbolivar replied:
> That is a minor problem because that is invalid syntax, but it would be great for
> usability if you were able to fix this!

Glad you agree it's OK to fix. It's fixed in my patch.

> I don't see how this affects the user, but if you can find a clean fix then why not :)

This script prints warnings on overrides. Warnings affect users.

mbolivar commented:
@nashif: there is good feedback from @SebastianBoe on my patch, which he left here:

https://github.com/mbolivar/zephyr/commit/e291dce056d0a9165bedf2ac1a06afda0ecd470b#commitcomment-26412231

I think we need you to decide if this should be loose or strict.

My preference is for merge_config.py to do some filtering to clean up the output and not raise false positives about overrides that are related to whitespace or comments.

nashif commented:
we should keep the behaviour as we have it now, both ways for setting a configuration option as 'n' should be supported IMO and they we do that right now in configurations (CONFIG_XYZ=n) should be kept, using the "is not set" is counter intuitive IMO, but still should be supported if someone has it in their conf.

@mbolivar  Can you submit your changes as a PR against my branch please, this would make it easier for me to pull it in once reviewed...

ulfalizer commented:
Hello!

I just stumbled upon this project via a notification. I'm the Kconfiglib guy.

In case you're interested, I hacked together a version of `merge_config.py` which offloads all the parsing and Kconfig arcana to Kconfiglib. It isn't thoroughly tested yet, but should give an idea of how it could be done.

```python
#!/usr/bin/env python3
from kconfiglib import Kconfig, Symbol, BOOL, TRISTATE, TRI_TO_STR
import sys

if len(sys.argv) < 4:
    print("usage: merge_config.py Kconfig merged_config config1 [config2 ...]")
    sys.exit(1)

kconf = Kconfig(sys.argv[1])

# Enable warnings for assignments to undefined symbols
kconf.enable_undef_warnings()

# (This script uses alldefconfig as the base. Other starting states could be
# set up here as well. The approach in examples/allnoconfig_simpler.py could
# provide an allnoconfig starting state for example.)

# Create a merged configuration by loading the fragments with replace=False
for config in sys.argv[3:]:
    kconf.load_config(config, replace=False)

# Write the merged configuration
kconf.write_config(sys.argv[2])

# Print warnings for symbols whose actual value doesn't match the assigned
# value
for sym in kconf.defined_syms:
    # Was the symbol assigned to?
    if sym.user_value is not None:
        # Tristate values are represented as 0, 1, 2. Having them as
        # "n", "m", "y" is more convenient here, so convert.
        if sym.type in (BOOL, TRISTATE):
            user_value = TRI_TO_STR[sym.user_value]
        else:
            user_value = sym.user_value

        if user_value != sym.str_value:
            print('warning: {} was assigned the value "{}" but got the '
                  'value "{}" -- check dependencies'
                  .format(sym.name, user_value, sym.str_value))
```

Here's a demo:

**Kconfig**
```
config FOO
	bool "FOO"

config BAR
	bool "BAR"

config BAZ
	string "BAZ"

config QAZ
	bool "QAZ" if n
```

**conf1**
```
CONFIG_FOO=y
```

**conf2**
```
CONFIG_BAR=y
```

**conf3**
```
# Ops... assigned twice (as a side note, CONFIG_FOO=n would work as well)
# CONFIG_FOO is not set

# Ops... this symbol doesn't exist
CONFIG_OPS=y

CONFIG_BAZ="baz string"
```

**conf4**
```
CONFIG_QAZ=y
```

**Running**
```
$ ./merge_config.py Kconfig merged conf1 conf2 conf3 conf4
conf3:2: warning: FOO set more than once. Old value: "y", new value: "n".
conf3:5: warning: attempt to assign the value "y" to the undefined symbol OPS
warning: QAZ was assigned the value "y" but got the value "n" -- check dependencies
$ cat merged
# Generated by Kconfiglib (https://github.com/ulfalizer/Kconfiglib)
# CONFIG_FOO is not set
CONFIG_BAR=y
CONFIG_BAZ="baz string"
```

Cheers,
Ulf

carlescufi commented:
If we're gonna use Kconfiglib to replace conf.exe with Python, then the solution proposed by @ulfalizer is as simple as it gets.

carlescufi commented:
@ulfalizer, could you let us know why your version of merge_config requires the actual Kconfig tree, and not just the list of fragment files to be merged?

SebastianBoe commented:
Very nice. Looks like this will make it easier to get rid of conf.

On Dec 29, 2017 5:34 PM, "Carles Cufí" <notifications@github.com> wrote:

> @ulfalizer <https://github.com/ulfalizer>, could you let us know why your
> version of merge_config requires the actual Kconfig tree, and not just the
> list of fragment files to be merged?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/5417#issuecomment-354468205>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAlLztbQNm_8E-HYjdrZ82iY_HuWdnsBks5tFRSBgaJpZM4REq4V>
> .
>

ulfalizer commented:
@carlescufi 
That matches my reading of `merge_config.sh` from the kernel tree. After all the config fragment files have been merged into `$TMP_FILE`, it runs the following command to fill out the generated `.config` with all the symbol values that haven't been specified. Those could come from defaults specified in a `Kconfig` file, for example.

```sh
make KCONFIG_ALLCONFIG=$TMP_FILE $OUTPUT_ARG $ALLTARGET
```

`$ALLTARGET` will be either `alldefconfig` or `allnoconfig`, and those require parsing the `Kconfig` files. Checking dependencies to make sure that the generated `.config` is well-formed also requires looking at them.

It's very possible that I've misunderstood some nuance of how it is supposed to work, so don't be afraid to correct me if I have. :)

carlescufi commented:
@ulfalizer The thing is that Zephyr uses it slightly differently than the kernel. Right now we invoke merge_config.sh like so:
```
merge_config.sh -m -q -O <dest path for .config> fragment1 fragment2 ...
```
Note the `-m` which disables the invocation of `make`. And then we run `conf.exe` separately passing the Kconfig tree path and the `.config` as parameters.

So if I understand you correctly your script does both the `merge_config.sh` and the actual `conf` run in one go, whereas we do it in 2 steps today. Our `.config` after running `merge_config.sh` is just a merge of all the fragments with a very small subset of default values, whereas with yours it would actually be the final `.config` used for building if I understand correctly.

ulfalizer commented:
@carlescufi 
Yep, right now it's generating the complete .config right away. That's probably the simplest approach if you can get away with it.

I actually did an earlier version that functions more like the Zephyr version before I looked closer at `merge_config.sh` though. Replacing the `kconf.write_config(sys.argv[2])` will the following gives that version:

```python
with open(sys.argv[2], "w") as merged_config:
    node = kconf.top_node

    while True:
        if isinstance(node.item, Symbol) and node.item.user_value is not None:
            if node.item.visibility:
                merged_config.write(node.item.config_string)
            else:
                print("warning: an assignment to {0} appeared in one of "
                      "the config fragments but couldn't be respected in "
                      "the merged configuration, due to {0} not being "
                      "visible".format(node.item.name))

        # Iterative tree walk using parent pointers

        if node.list:
            node = node.list
        elif node.next:
            node = node.next
        else:
            while node.parent:
                node = node.parent
                if node.next:
                    node = node.next
                    break
            else:
                break
```

The iterative tree walk could be replaced with something simpler to read like in `examples/print_tree.py`. It was just handy to copy it out of the `write_config()` implementation.

That spammy warning could be replaced with something more informative like in the first version too.

carlescufi commented:
@ulfalizer thanks very much for all the suggestions.
@mbolivar @nashif @SebastianBoe: Actually and given all the input here, I would be in favor of using the `merge_config.py` that @ulfalizer has posted [here](https://github.com/zephyrproject-rtos/zephyr/pull/5417#issuecomment-354465512) and call it something like `kconfig.py` instead, since it does everything in one go.

ulfalizer commented:
That version still looks at the `Kconfig` tree though. Kconfiglib unfortunately isn't well-suited to processing `.config` files in isolation.

carlescufi commented:
@ulfalizer yes, and that's why I think the better solution is to consolidate the Kconfig parsing and the fragment merging in your first version instead of keeping it separate.

ulfalizer commented:
@carlescufi Yeah, I think that might lead to less head-scratching from tricky dependency stuff in the end too.

carlescufi commented:
@ulfalizer since both you and @nashif are here, I think there was a change that we required to `Kconfiglib` to be able to use it properly, which is "glob" support as shown here: https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/Kconfig#L19. Is this something we could consider sending a PR for?

ulfalizer commented:
@carlescufi 
I'd be happy to add it. Don't think it should interfere with vanilla Kconfig parsing.

A loop in the `_T_SOURCE` handling ought to do it.

carlescufi commented:
@ulfalizer well thanks very much for the offer! I will let @nashif confirm since he's the one that has been looking into this switch from C to Python in detail.

nashif commented:
> A loop in the _T_SOURCE handling ought to do it.

already have a patch for this:

```patch
commit 35c4ef1f65d37c4e0cd0c6fea9fa9138a9418eb0
Author: Anas Nashif <anas.nashif@intel.com>
Date:   Tue Dec 12 10:46:23 2017 -0500

    zephyr: support globbing

    Signed-off-by: Anas Nashif <anas.nashif@intel.com>

diff --git a/kconfiglib.py b/kconfiglib.py
index 4d763a3..254c8b4 100644
--- a/kconfiglib.py
+++ b/kconfiglib.py
@@ -368,6 +368,7 @@ import os
 import platform
 import re
 import sys
+import glob

 # File layout:
 #
@@ -1453,12 +1454,15 @@ class Kconfig(object):
                 prev_node.next = prev_node = node

             elif t0 == _T_SOURCE:
-                self._enter_file(self._expand_syms(self._next_token()))
-                prev_node = self._parse_block(None,            # end_token
-                                              parent,
-                                              visible_if_deps,
-                                              prev_node)
-                self._leave_file()
+                f = self._expand_syms(self._next_token())
+                g = glob.glob(f)
+                for s in g:
+                    self._enter_file(s)
+                    prev_node = self._parse_block(None,            # end_token
+                                                  parent,
+                                                  visible_if_deps,
+                                                  prev_node)
+                    self._leave_file()

             elif t0 == end_token:
                 # We have reached the end of the block. Terminate the final
```

ulfalizer commented:
Yeah, that's the reasonable approach. One thing that gets a bit icky is that globbing also ought to be done relative to `$srctree` if set, so that out-of-tree configuration works. Do you ever do those?

Small nitpick: The imports are sorted alphabetically.

If you submit a PR, we could pick up there.

ulfalizer commented:
Unless there's a huge number of `Kconfig` files or some dynamic stuff going on with the tree, it seems simpler to just list them all explicitly though. Easier to selectively disable individual `Kconfig` files for debugging then too.

nashif commented:
> Yeah, that's the reasonable approach. One thing that gets a bit icky is that globbing also ought to be done relative to $srctree if set, so that out-of-tree configuration works. Do you ever do those?
> 
> Small nitpick: The imports are sorted alphabetically.
> 
> If you submit a PR, we could pick up there.

will do after some more testing, thanks.

ulfalizer commented:
If you need some tests, something like this in `testsuite.py` should do, if `Kglobbed1-3` are globbed up in various ways at the end of `tests/Klocation` (so that nothing else moves and breaks the existing tests) and each contain a single symbol.

```diff
+    # Test 'source' globbing
+
+    verify_locations(c.syms["GLOBBED1"].nodes, "tests/Kglobbed1:1")
+    verify_locations(c.syms["GLOBBED2"].nodes, "tests/Kglobbed2:1")
+    verify_locations(c.syms["GLOBBED3"].nodes, "tests/sub/Kglobbed3:1")
+
 
     print("Testing visibility")
```

carlescufi commented:
@ulfalizer I've posted a PR with globbing support including $srctree in your GitHub repo. However when running your proposed script [here](https://github.com/zephyrproject-rtos/zephyr/pull/5417#issuecomment-354465512), and although the configuration is parsed correctly, the .defconfig and .conf files seem not to apply, and the architecture is set to the default x86 even though I'm providing a defconfig that sets it to ARM. Any ideas?

carlescufi commented:
@ulfalizer how hard would it be to write a kconfigdiff with your library? something that is able to load 2 `.config` files and compare them regardless of the order the variables appear? This would be useful actually to make sure the transition from conf to Kconfiglib has not introduced changes.

carlescufi commented:
@ulfalizer So after playing around a bit more with your script with globbing added, I see a few differences between the `.config` generated by `conf` and the one generated by Kconfiglib. Those revolve mainly around items that default to a value with an `if` statement such as:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/Kconfig#L29

In this case, and for my particular application, `BT_RX_BUF_COUNT` should default to 3 because `BT_RECV_IS_RX_THREAD` is enabled, but in the `.config` generated by Kconfiglib I get 10 instead.

ulfalizer commented:
@carlescufi 
Do you think your `conf` utility might be out of sync with the kernel's?

Normally, `conf` walks through the defaults in order, picking the first one with a satisfied condition. A missing condition, like for `default 10`, is always satisfied, so any default after it shouldn't even be considered. The defaults would need to be swapped.

Here's the code from the C implementation (in `scripts/kconfig/symbol.c`), where `visible` is the condition:

```C
static struct property *sym_get_default_prop(struct symbol *sym)
{
	struct property *prop;

	for_all_defaults(sym, prop) {
		prop->visible.tri = expr_calc_value(prop->visible.expr);
		if (prop->visible.tri != no)
			return prop;
	}
	return NULL;
}
```
Here's the corresponding Kconfiglib code (which also includes some logic from `sym_calc_value()` in the C implementation):

```python
for default, cond in self.defaults:
    cond_val = expr_value(cond)
    if cond_val:
        val = min(expr_value(default), cond_val)
        self._write_to_conf = True
        break
```

I suspect some things would fail in Kconfiglib's compatibility test suite too if the behavior didn't match here.

Here's two test files you could use (with e.g. `$ scripts/kconfig/conf default`) to see this behavior:

**default**
```
config FOO
	bool "foo"

config INT
	int "int"
	default 10
	default 12 if FOO
```

**.config**
```
CONFIG_FOO=y
```

ulfalizer commented:
@carlescufi 
One way to compare two `.config` files would be to create two `Kconfig` instances, load the two `.config` files into them, and then go through all symbols in both of them and compare the ones that have a `Symbol.user_value` that is not `None` (a similar thing is going on in the `merge_config.py` example too, so it might be good reference).

The `Symbol.user_value` field holds the value that was assigned to the symbol in the `.config` file. This value might be different from `Symbol.str_/tri_value`, because Kconfig always respects dependencies to avoid creating invalid configurations.

carlescufi commented:
@ulfalizer 

> Do you think your conf utility might be out of sync with the kernel's?

Yes, that is possible, although I am surprised about such a fundamental difference 

> Normally, conf walks through the defaults in order, picking the first one with a satisfied condition

So what you are saying is that in Linux's `Kconfig` files, the default with no condition comes always as the last one after all the ones with conditions? That's definitely not the case with Zephyr.

> Here's two test files you could use (with e.g. $ scripts/kconfig/conf default) to see this behavior:

I assume here Linux's `conf` and Kconfiglib would default `INT` to 10, but you expect Zephyr's `conf` to default it to 12, is that correct?

ulfalizer commented:
@carlescufi 
Yep. `INT` defaults to 10 with `conf`/`mconf` with the Linux Kconfig tools.

Someone changing the behavior there seems a bit arbitrary, but it's the only explanation I can think of. How heavily modified (and old) is the Zephyr `conf`?

Here's the kind of stuff you see in the Linux Kconfig files:

```
config PGTABLE_LEVELS
	int
	default 5 if X86_5LEVEL
	default 4 if X86_64
	default 3 if X86_PAE
	default 2
```

mbolivar commented:
Happy new year all :).

> @mbolivar @nashif @SebastianBoe: Actually and given all the input here, I would be in favor of using the merge_config.py that @ulfalizer has posted here and call it something like kconfig.py instead, since it does everything in one go.

+1 to not reinventing the wheel.

carlescufi commented:
@mbolivar I have Zephyr building with Kconfiglib, but unfortunately not all the changes can go upstream to @ulfalizer's repo due to differences between Linux's Kconfig and ours. autoconf generation can though, and it's currently in PR form: https://github.com/ulfalizer/Kconfiglib/pull/35

mbolivar commented:
> @mbolivar I have Zephyr building with Kconfiglib, but unfortunately not all the changes can go upstream to @ulfalizer's repo due to differences between Linux's Kconfig and ours.

@carlescufi saw the discussion about `default` ordering -- that's what you're referring to, right? Are you saying that we should fix/keep our own merge_config.py and incompatible `default` order in our Kconfigs?

I assumed the way forward was to fix the Zephyr Kconfigs so they are compatible with Linux, so we can use kconfiglib and the "standard" conf. Sorry if I misunderstood.

Happy to pick up my changes to merge_config.py if we're going to keep our incompatibilities for the time being.

carlescufi commented:
@mbolivar happy new year!
> @carlescufi saw the discussion about default ordering -- that's what you're referring to, right? Are you 
> saying that we should fix/keep our own merge_config.py and incompatible default order in our 
> Kconfigs?

There are 2 different items that make our Kconfig files incompatible with vanilla Linux Kconfig:

* Globbing: We require support for globbing when sourcing Kconfig files
* Prefer later defaults: Zephyr "prefers" the later defaults instead of matching the first one 

I plan to do this in 2 steps: First leave the Kconfig files unchanged but use a modified version of Kconfiglib to reproduce our current behavior. This is to ensure that Kconfiglib indeed respects everything else about Kconfig and to avoid changes in Kconfig files at the same time that we replace `conf`. Then in a second step I suggest we remove the globbing (it's silly anyway IMHO) and that we go back to standard default matching.

> Happy to pick up my changes to merge_config.py if we're going to keep our incompatibilities for the 
> time being.

There's no more need for a merge_config.py. As soon as I post my PR you'll see that there's a single `kconfig.py` that does both merge_config and `conf` in one go, as well as writing the `autoconf.h`.

carlescufi commented:
I think we should close this PR now that https://github.com/zephyrproject-rtos/zephyr/pull/5569 is posted.
